Source/WebGPU/WGSL/TypeStore.cpp

```diff
@@ -67,27 +105,41 @@ Type* TypeStore::structType(const AST::Identifier& name)
 
 Type* TypeStore::constructType(AST::ParameterizedTypeName::Base base, Type* elementType)
 {
-    // FIXME: these should be cached
     auto& typeConstructor = m_typeConstrutors[WTF::enumToUnderlyingType(base)];
     return typeConstructor.construct(elementType);
 }
 
 Type* TypeStore::arrayType(Type* elementType, std::optional<unsigned> size)
 {
-    // FIXME: these should be cached
-    return allocateType<Array>(elementType, size);
+    ArrayKey key { elementType, size };
+    Type* type = m_cache.find(key);
+    if (type)
+        return type;
+    type = allocateType<Array>(elementType, size);
+    m_cache.insert(key, type);
+    return type;
 }
 
 Type* TypeStore::vectorType(Type* elementType, uint8_t size)
 {
-    // FIXME: these should be cached
-    return allocateType<Vector>(elementType, size);
+    VectorKey key { elementType, size };
+    Type* type = m_cache.find(key);
+    if (type)
+        return type;
+    type = allocateType<Vector>(elementType, size);
+    m_cache.insert(key, type);
+    return type;
 }
 
 Type* TypeStore::matrixType(Type* elementType, uint8_t columns, uint8_t rows)
 {
-    // FIXME: these should be cached
-    return allocateType<Matrix>(elementType, columns, rows);
+    MatrixKey key { elementType, columns, rows };
+    Type* type = m_cache.find(key);
+    if (type)
+        return type;
+    type = allocateType<Matrix>(elementType, columns, rows);
```

Comment:
Shouldn't we just preallocate all the matrix types? There's only a limit number of vector and matrix types.

Reply:
We could, but that'd require 27 matrix types ahead of time, which seems unnecessary? I think very few programs will use them all